File: linkedlist.rs
#[derive(Debug)]
struct Link {
    val: Option<i32>,
    next: Option<Box<Link>>,
}

impl Link {
    fn new(val: i32) -> Self {
        Self {
            val: Some(val),
            next: None,
        }
    }

    fn new_dummy() -> Self {
        Self {
            val: None,
            next: None,
        }
    }
}

#[derive(Debug)]
struct List {
    head: Option<Box<Link>>,
    curr: Option<*mut Link>,
}

impl List {
    fn new() -> Self {
        let dummy = Box::new(Link::new_dummy());
        let raw_dummy = Box::into_raw(dummy);

        Self {
            head: unsafe { Some(Box::from_raw(raw_dummy)) },
            curr: Some(raw_dummy),
        }
    }

    fn append(&mut self, val: i32) {
        let new_link = Box::new(Link::new(val));
        let raw_new_link = Box::into_raw(new_link);

        if let Some(curr) = self.curr {
            unsafe {
                (*curr).next = Some(Box::from_raw(raw_new_link));
            }
        }
        self.curr = Some(raw_new_link);
    }
    fn print_all(&self) {}
}

fn main() {
    let mut list = List::new();
    list.append(10);
    list.append(20);
    list.append(30);

    println!("{:?}", list.head);
}

File: linkedlist.cpp
#include <bits/stdc++.h>
using namespace std;

class Link {
public:
  int data;
  Link *next;
  Link(int data) {
    this->data = data;
    this->next = nullptr;
  }
};

class LL {
  Link *head;
  Link *last;

public:
  LL() { this->head = this->last = nullptr; }
  void add(int data) {
    Link *new_link = new Link(data);
    new_link->next = head;
    head = new_link;
  }
  void print() {
    cout << "Showing Nodes:" << endl;
    Link *curr = head;
    while (curr != nullptr) {
      cout << curr->data << endl;
      curr = curr->next;
    }
  }
};

int main() {
  LL ll;
  ll.print();
  ll.add(5);
  ll.print();
  ll.add(10);
  ll.print();
  ll.add(15);
  ll.print();
}
